In [1]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [61]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [7]:
porcentaje_vacios = ((wh.isnull().sum() / len(wh))*100).sort_values(ascending = False)

In [8]:
porcentaje_vacios

Healthy life expectancy         8.333333
Freedom to make life choices    7.371795
Generosity                      6.730769
Score                           6.089744
Social support                  5.128205
Perceptions of corruption       4.166667
GDP per capita                  3.846154
Overall rank                    0.000000
Country or region               0.000000
year                            0.000000
dtype: float64

In [48]:
x = pd.DataFrame({
    "porcentaje vacios": porcentaje_vacios
})

In [49]:
x

porcentaje vacios                      Columnas
Healthy life expectancy                8.333333                  Overall rank
Freedom to make life choices           7.371795             Country or region
Generosity                             6.730769                         Score
Score                                  6.089744                GDP per capita
Social support                         5.128205                Social support
Perceptions of corruption              4.166667       Healthy life expectancy
GDP per capita                         3.846154  Freedom to make life choices
Overall rank                           0.000000                    Generosity
Country or region                      0.000000     Perceptions of corruption
year                                   0.000000                          year

In [67]:
problema1 = wh.groupby( ["Overall rank"], as_index = False )

In [68]:
problema1 =  problema1.loc[:,["Overall rank"]].groupby(["Overall rank"], as_index = False)

AttributeError: 'DataFrameGroupBy' object has no attribute 'loc'

In [78]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
3               NaN                    0.914                         0.677   
4             1.549                    0.927                         0.660   
..              ...                      ...                           ...   
151           0.711                    0.614                         0.555   
152           0.885                    0.499                         0.417   
153           0.517                    0.361                         0.000   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0         0.202                      0.393  None  
1         0.286                      0.340  None  
2         0.284                      0.408  None  
3           NaN                      0.138  None  
4         0.256                      0.357  None  
..          ...                        ...   ...  
151       0.217                      0.411  None  
152       0.276                      0.147  None  
153       0.158                      0.025  None  
154       0.235                      0.035  None  
155       0.202                      0.091  None  

[312 rows x 10 columns]

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [75]:
wh.year = wh.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 312 entries, 0 to 155
Data columns (total 10 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  312 non-null    int64  
 1   Country or region             312 non-null    object 
 2   Score                         293 non-null    float64
 3   GDP per capita                300 non-null    float64
 4   Social support                296 non-null    float64
 5   Healthy life expectancy       286 non-null    float64
 6   Freedom to make life choices  289 non-null    float64
 7   Generosity                    291 non-null    float64
 8   Perceptions of corruption     299 non-null    float64
 9   year                          312 non-null    int64  
dtypes: float64(7), int64(2), object(1)
memory usage: 34.9+ KB


In [122]:
wh.describe()

Overall rank       Score  GDP per capita  Social support  \
count    312.000000  293.000000      300.000000      296.000000   
mean      78.500000    5.392816        0.896833        1.207145   
std       45.104737    1.108112        0.399329        0.302437   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.519000        0.588500        1.047750   
50%       78.500000    5.386000        0.954000        1.265500   
75%      117.250000    6.182000        1.221500        1.455500   
max      156.000000    7.769000        2.096000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               286.000000                    289.000000  291.000000   
mean                  0.659832                      0.425872    0.183309   
std                   0.256373                      0.155103    0.096846   
min                   0.000000                      0.000000    0.000000   
25%                   0.485250                      0.324000    0.110000   
50%                   0.695500                      0.450000    0.175000   
75%                   0.859750                      0.543000    0.244000   
max                   1.141000                      0.724000    0.598000   

       Perceptions of corruption  
count                 299.000000  
mean                    0.110950  
std                     0.095574  
min                     0.000000  
25%                     0.050000  
50%                     0.082000  
75%                     0.139500  
max                     0.457000

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [112]:
a1 = wh['Score'].mean()

In [107]:
a2 = wh['GDP per capita'].mean()

In [89]:
a3 = wh['Social support'].mean()

In [90]:
a4 = wh['Healthy life expectancy'].mean()

In [91]:
a5 = wh['Freedom to make life choices'].mean()

In [92]:
a6 = wh['Generosity'].mean()

In [ ]:
a7 = wh['Perceptions of corruption'].mean()

In [99]:
wh1 = wh.groupby(['year']).mean()

In [115]:
problema3 = wh.groupby( ["Overall rank","Country or region"] , as_index = False ).agg({"year":["count"]})

In [117]:
problema3.columns = problema3.columns.droplevel(1)

In [121]:
wh.fillna(0).mean

<bound method NDFrame._add_numeric_operations.<locals>.mean of      Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.59

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro